File: class 3/1074.py


import sys
n, r, c = map(int, sys.stdin.readline().split())
result = 0


while n != 0 :
    n -= 1

    if r < 2** n  and c < 2 ** n : # 1사분면
        result += (2 ** n) * (2 ** n ) * 0

    elif r < 2** n  and c >= 2 ** n : # 2사분면 
        result += (2 ** n) * (2 ** n ) * 1
        c -= (2 ** n)

    elif r >= 2** n  and c < 2 ** n : # 3사분면 
        result += (2 ** n) * (2 ** n ) * 2
        r -= (2**n)

    else : # 4사분면 
        result += (2 ** n) * (2 ** n ) * 3
        c -= (2 ** n)
        r -= (2**n)

print(result)

File: 백준 문제 풀이 모음/kotlin/class 3/1074.kt
package class3

var count = 0
var result = 0

fun main () = with(System.`in`.bufferedReader()) {
    val str = readLine().split(' ').toList()
    val n = str[0].toInt()
    val r = str[1].toInt()
    val c = str[2].toInt()

    divide(1 shl(n), 0, 0, r, c)

    println(result)
}

fun divide(size: Int, row: Int, col: Int, r: Int, c: Int) {
    if (r !in row until row + size || c !in col until col + size) {
        count += (size * size)
        return
    }

    if (r == row && c == col) {
        result = count
        return
    }

    divide(size/2, row, col, r, c)
    divide(size/2, row, col + size/2, r, c)
    divide(size/2, row + size/2, col, r, c)
    divide(size/2, row + size/2, col + size/2, r, c)






}
